rules/terraformrules/terraform_unused_required_providers.go

```diff
@@ -0,0 +1,79 @@
+package terraformrules
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/terraform/configs"
+	"github.com/terraform-linters/tflint/tflint"
+)
+
+// TerraformUnusedRequiredProvidersRule checks whether required providers are used in the module
+type TerraformUnusedRequiredProvidersRule struct{}
+
+// NewTerraformUnusedRequiredProvidersRule returns new rule with default attributes
+func NewTerraformUnusedRequiredProvidersRule() *TerraformUnusedRequiredProvidersRule {
+	return &TerraformUnusedRequiredProvidersRule{}
+}
+
+// Name returns the rule name
+func (r *TerraformUnusedRequiredProvidersRule) Name() string {
+	return "terraform_unused_required_providers"
+}
+
+// Enabled returns whether the rule is enabled by default
+func (r *TerraformUnusedRequiredProvidersRule) Enabled() bool {
+	return false
+}
+
+// Severity returns the rule severity
+func (r *TerraformUnusedRequiredProvidersRule) Severity() string {
+	return tflint.WARNING
+}
+
+// Link returns the rule reference link
+func (r *TerraformUnusedRequiredProvidersRule) Link() string {
+	return tflint.ReferenceLink(r.Name())
+}
+
+//Check checks whether required providers are used
+func (r *TerraformUnusedRequiredProvidersRule) Check(runner *tflint.Runner) error {
+	if !runner.TFConfig.Path.IsRoot() {
+		// This rule does not evaluate child modules.
+		return nil
+	}
+
+	log.Printf("[TRACE] Check `%s` rule for `%s` runner", r.Name(), runner.TFConfigPath())
+
+	for _, required := range runner.TFConfig.Module.ProviderRequirements.RequiredProviders {
+		r.checkProvider(runner, required)
+	}
+
+	return nil
+}
+
+func (r *TerraformUnusedRequiredProvidersRule) checkProvider(runner *tflint.Runner, required *configs.RequiredProvider) {
+	for _, resource := range runner.TFConfig.Module.ManagedResources {
+		if required.Name == resource.Provider.Type {
```

Comment:
If you don't use a preferred local name, this condition may not work as expected.
https://www.terraform.io/docs/configuration/provider-requirements.html#local-names

Example:

```hcl
terraform {
  required_providers {
    mycloud = {
      source = "hashicorp/aws"
      version = "3.20.0"
    }
  }
}

resource "aws_instance" "foo" {
  provider = mycloud

  ami = "ami-12345678"
  instance_type = "t2.micro"
}
```

Reply:
Good catch, handled in 3a3b486